### PR TITLE
Add retrieval artifact extraction, visualizations, and demographic analysis

### PR DESCRIPTION
**Pull request by zaczw**
## Summary
- Add `predict_step` to `MedRAPSupervisedLightningModule` that collects per-sample retrieval artifacts (doc IDs, scores, query embeddings, key embeddings, per-doc logits, targets)
- Add `extraction.py` with `collate_prediction_batches()` and `extract_artifacts()` for running extraction and saving to `.pt`
- Add `demographic_analysis.py` for retrieval-behavior breakdowns by patient demographics: age binning, race aggregation, softmax helpers, title-keyword / static-mapping / LDA topic doc-keyword providers, keyword × demographic tables, subject demographic loading, and `render_demographic_heatmaps` for stacked age / race / gender PNG output
- Add `scripts/extract_and_visualize.py` that loads a trained run, extracts artifacts, and generates a 6-panel figure characterizing retrieval behavior (query embedding structure, doc key embeddings, score-prediction alignment, class-conditionality of soft retrieval)
- Add `scripts/run_demographic_heatmap.py` and supporting debug scripts (`debug_artifacts.py`, `debug_gender_alignment.py`, `debug_score_variance.py`, `debug_vocab.py`) for per-run demographic diagnostics
- Add `train_embedding_ids_check.sh` 
- Declare `matplotlib` as an optional `viz` extra and wire it into `tests.yaml` so matplotlib-dependent tests can run in CI
- Minor updates to `retrievers.py`, `.gitignore`, `.pre-commit-config.yaml`, and `pyproject.toml`

## Tests
- `tests/test_extraction.py` — predict_step keys/shapes, batch collation, missing keys, empty input, full extract roundtrip
- `tests/test_lightning_module.py` — `predict_step` target extraction, graceful fallback when `extract_targets` raises, and capture of marginalized tensors (`per_doc_logits`, `differentiable_doc_scores`) from `ModelOutput.metadata`
- `tests/test_demographic_analysis.py` — age/race binning, softmax stability, title-keyword / static-mapping / LDA topic providers, keyword × demographic table shape / normalization / trimming / multi-topic weighting / input validation, doc-id row-mapping resolution, validation-schema extraction, subject demographic loading, and `render_demographic_heatmaps` PNG output + input validation
- Doctests across new/updated modules

### CODE REVIEW

File: scripts/extract_and_visualize.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+    # ------------------------------------------------------------------
+    ax4 = axes[1, 0]
+    if has_per_doc:
+        c_idx = min(1, per_doc_logits.shape[-1] - 1)
+        n_patients, k_docs = diff_scores.shape
+
+        # x: each doc's positive logit (per patient-doc pair)
+        x_vals = per_doc_logits[:, :, c_idx].reshape(-1)
+        # y: per-doc positive logit (what this specific doc predicts for this patient)
+        y_vals = per_doc_logits[:, :, c_idx].reshape(-1)
```
Comment (stepanyanhayk) [inline scripts/extract_and_visualize.py]: x_vals and y_vals are the same here; is this intended?

File: scripts/extract_and_visualize.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+        accelerator="cpu",
+        devices=1,
+        logger=False,
+        enable_checkpointing=False,
+        enable_progress_bar=False,
+    )
+
+    # Grab raw code values from the first batch (for scatter plot).
+    first_batch = next(iter(dataloader))
+    code_means: Tensor | None = None
```
Comment (stepanyanhayk) [inline scripts/extract_and_visualize.py]: is this used?

**Review by stepanyanhayk — APPROVED** [review]

File: scripts/extract_and_visualize.py (hunk 1 above)
Comment (zaczw) [reply scripts/extract_and_visualize.py]: Changed. Originally was doing a sanity check, but now modified to a substantive plot.

File: scripts/extract_and_visualize.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+
+    # Grab full corpus key embeddings from the retriever (if InMemoryRetriever).
+    all_doc_keys: Tensor | None = None
+    retriever = getattr(module.model, "retriever", None)
+    if retriever is not None and hasattr(retriever, "_doc_key_embeddings"):
+        all_doc_keys = retriever._doc_key_embeddings.detach().clone()
+
+    # Datamodule — val split (deterministic, no shuffle).
+    datamodule = instantiate_datamodule(cfg)
+    dataloader: DataLoader = datamodule.val_dataloader()
```
[the remark this answers is not in the document]
Comment (zaczw) [reply scripts/extract_and_visualize.py:83]: Added.

File: scripts/extract_and_visualize.py (hunk 2 above)
Comment (zaczw) [reply scripts/extract_and_visualize.py]: Removed, was legacy code.

File: scripts/run_retrieval_only.sh
Comment (stepanyanhayk) [inline scripts/run_retrieval_only.sh:1]: Please don't change this file. I updated it in mimic-model-training with correct params

File: scripts/train_wandb_slurm.sh
Comment (stepanyanhayk) [inline scripts/train_wandb_slurm.sh:1]: i don't think we need this file anymore. It was created for testing purposes then I removed it from my branch
